modules/pubmaticBidAdapter.js

```diff
@@ -2,137 +2,216 @@ var utils = require('src/utils.js');
+    iframeDoc.write(content);
     iframeDoc.close();
   }
 
-  function _createRequestContent() {
-    var content = '<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"' +
-      ' "http://www.w3.org/TR/html4/loose.dtd"><html><head><base target="_top" /><scr' +
-      'ipt>inDapIF=true;</scr' + 'ipt></head>';
-    content += '<body>';
-    content += '<scr' + 'ipt>';
-    content += '' +
-      'window.pm_pub_id  = "%%PM_PUB_ID%%";' +
-      'window.pm_optimize_adslots     = [%%PM_OPTIMIZE_ADSLOTS%%];' +
-      'window.kaddctr = "%%PM_ADDCTR%%";' +
-      'window.kadgender = "%%PM_GENDER%%";' +
-      'window.kadage = "%%PM_AGE%%";' +
-      'window.pm_async_callback_fn = "window.parent.$$PREBID_GLOBAL$$.handlePubmaticCallback";';
-
-    content += '</scr' + 'ipt>';
-
-    var map = {};
-    map.PM_PUB_ID = _pm_pub_id;
-    map.PM_ADDCTR = _pm_pub_kvs;
-    map.PM_GENDER = _pm_pub_gender;
-    map.PM_AGE = _pm_pub_age;
-    map.PM_OPTIMIZE_ADSLOTS = _pm_optimize_adslots.map(function (adSlot) {
-      return "'" + adSlot + "'";
-    }).join(',');
-
-    content += '<scr' + 'ipt src="https://ads.pubmatic.com/AdServer/js/gshowad.js"></scr' + 'ipt>';
-    content += '<scr' + 'ipt>';
-    content += '</scr' + 'ipt>';
-    content += '</body></html>';
-    content = utils.replaceTokenInString(content, map, '%%');
-
-    return content;
+  function _generateLegacyCall(conf, slots) {
+    var request_url = 'gads.pubmatic.com/AdServer/AdCallAggregator';
+    return _protocol + request_url + '?' + utils.parseQueryStringParameters(conf) + 'adslots=' + encodeURIComponent('[' + slots.join(',') + ']');
   }
 
-  $$PREBID_GLOBAL$$.handlePubmaticCallback = function () {
-    let bidDetailsMap = {};
-    let progKeyValueMap = {};
-    try {
-      bidDetailsMap = iframe.contentWindow.bidDetailsMap;
-      progKeyValueMap = iframe.contentWindow.progKeyValueMap;
-    } catch (e) {
-      utils.logError(e, 'Error parsing Pubmatic response');
+  function _initUserSync(pubId) {
+    // istanbul ignore else
+    if (!usersync) {
+      var iframe = utils.createInvisibleIframe();
+      iframe.src = _protocol + 'ads.pubmatic.com/AdServer/js/showad.js#PIX&kdntuid=1&p=' + pubId;
+      utils.insertElement(iframe, document);
+      usersync = true;
+    }
+  }
+
+  function _callBids(params) {
+    var conf = _initConf();
+    var slots = [];
+
+    conf.pubId = 0;
+    bids = params.bids || [];
+
+    for (var i = 0; i < bids.length; i++) {
+      var bid = bids[i];
+      conf.pubId = conf.pubId || bid.params.publisherId;
+      conf = _handleCustomParams(bid.params, conf);
+      bid.params.adSlot = _cleanSlot(bid.params.adSlot);
+      bid.params.adSlot.length && slots.push(bid.params.adSlot);
+    }
+
+    // istanbul ignore else
+    if (conf.pubId && slots.length > 0) {
+      _legacyExecution(conf, slots);
```

Comment:
it looks like the legacy path is currently the only supported path (ie it still uses an iframe) is that correct? Is this optimized vs the last version?

Reply:
Our older implementation used to include an external js library, now we do not include it.
_legacyExecution is the only flow available now, we are planning to add a new flow later.